src/Patterns/Patterns.php

```diff
@@ -0,0 +1,193 @@
+	 *
+	 *     /**
+	 *      * Title: My Pattern
+	 *      * Slug: my-theme/my-pattern
+	 *      *
+	 *
+	 * The output of the PHP source corresponds to the content of the pattern, e.g.:
+	 *
+	 *     <main><p><?php echo "Hello"; ?></p></main>
+	 *
+	 * Other settable fields include:
+	 *
+	 *   - Description
+	 *   - Viewport Width
+	 *   - Categories       (comma-separated values)
+	 *   - Keywords         (comma-separated values)
+	 *   - Block Types      (comma-separated values)
+	 *   - Inserter         (yes/no)
+	 *
+	 * @since 6.0.0
+	 */
+	public function register_block_patterns() {
+		if ( ! class_exists( 'WP_Block_Patterns_Registry' ) ) {
+			return;
+		}
+
+		$default_headers = array(
+			'title'         => 'Title',
+			'slug'          => 'Slug',
+			'description'   => 'Description',
+			'viewportWidth' => 'Viewport Width',
+			'categories'    => 'Categories',
+			'keywords'      => 'Keywords',
+			'blockTypes'    => 'Block Types',
+			'inserter'      => 'Inserter',
+		);
+
+		if ( file_exists( $this->patterns_path ) ) {
+			$files = glob( $this->patterns_path . '/*.php' );
+			if ( $files ) {
+				foreach ( $files as $file ) {
+					$pattern_data = get_file_data( $file, $default_headers );
+
+					if ( empty( $pattern_data['slug'] ) ) {
+						_doing_it_wrong(
+							'register_block_patterns',
+							esc_html(
+								sprintf(
+								/* translators: %s: file name. */
+									__( 'Could not register file "%s" as a block pattern ("Slug" field missing)', 'woo-gutenberg-products-block' ),
+									$file
+								)
+							),
+							'6.0.0'
+						);
+						continue;
+					}
+
+					if ( ! preg_match( '/^[A-z0-9\/_-]+$/', $pattern_data['slug'] ) ) {
+						_doing_it_wrong(
+							'register_block_patterns',
+							esc_html(
+								sprintf(
+								/* translators: %1s: file name; %2s: slug value found. */
+									__( 'Could not register file "%1$s" as a block pattern (invalid slug "%2$s")', 'woo-gutenberg-products-block' ),
+									$file,
+									$pattern_data['slug']
+								)
+							),
+							'6.0.0'
+						);
+					}
+
+					if ( \WP_Block_Patterns_Registry::get_instance()->is_registered( $pattern_data['slug'] ) ) {
+						continue;
+					}
+
+					// Title is a required property.
+					if ( ! $pattern_data['title'] ) {
+						_doing_it_wrong(
+							'register_block_patterns',
+							esc_html(
+								sprintf(
+								/* translators: %1s: file name; %2s: slug value found. */
+									__( 'Could not register file "%s" as a block pattern ("Title" field missing)', 'woo-gutenberg-products-block' ),
+									$file
+								)
+							),
+							'6.0.0'
+						);
+						continue;
+					}
+
+					// For properties of type array, parse data as comma-separated.
+					foreach ( array( 'categories', 'keywords', 'blockTypes' ) as $property ) {
+						if ( ! empty( $pattern_data[ $property ] ) ) {
+							$pattern_data[ $property ] = array_filter(
+								preg_split(
+									'/[\s,]+/',
```

Comment:
Can we comment this regex, or move it to a named variable so it's more descriptive?